### PR TITLE
Timeout Unschedulable Migration Target Pods

### PR DESCRIPTION
**Pull request by davidvossel**
This solves the issue where migrations can get stuck indefinitely when a migration target pod is unschedulable.

In the future, we may want to expand this logic to include any other situations we encounter where a migration can get stuck indefinitely as well.

```release-note
Pending migration target pods timeout after 5 minutes when unschedulable
```

### CODE REVIEW

**Comment by davidvossel** [comment]
/hold 

I want to write a functional test before this is merged.

File: pkg/virt-controller/watch/migration.go
Comment (maiqueb) [inline pkg/virt-controller/watch/migration.go:844]: Maybe I'm just too thick, but I don't follow this `c.unschedulableTimeoutSeconds-diffSeconds` . 

Would you assign it to a named variable to help people like me reason this out ?

File: pkg/virt-controller/watch/migration.go
Comment (maiqueb) [inline pkg/virt-controller/watch/migration.go:805]: I don't think this variable is required, nor the control code to exit the function if the pod is not `k8sv1.PodReasonUnschedulable` .

You could move all the code in https://github.com/kubevirt/kubevirt/pull/6737/files#diff-1baf192bb9c4560d7d40f008b7ba5c680ebce669b8f1b7639445abd3e7269688R820-R844 to a couple of new functions (named `didPodTimeOut(pod *v1.Pod) bool {...}` and `func (c *MigrationController) podTimedOut(vmi, pod) error {...}`) then directly call them when the right condition is met - i.e. `condition.Type == k8sv1.PodScheduled` && `condition.Status == k8sv1.ConditionFalse` && `condition.Reason == k8sv1.PodReasonUnschedulable`.

Not a must though, I won't insist.

File: pkg/virt-controller/watch/migration_test.go
Comment (maiqueb) [inline pkg/virt-controller/watch/migration_test.go:603]: I'm not fond of defining the expected test state in the test's input parameters - i.e. the `shouldTimeout` boolean input argument - but I don't see simpler alternatives.

**Review by maiqueb — COMMENTED** [review]
Just a couple of comments; the most important one - imo - is to name the variable to shed some light into the re-queing mechanism.

**Comment by davidvossel** [comment]
/hold cancel

I added a functional test

File: pkg/virt-controller/watch/migration.go
Comment (davidvossel) [reply pkg/virt-controller/watch/migration.go]: i changed this around a bit to try and help

File: pkg/virt-controller/watch/migration.go
Comment (davidvossel) [reply pkg/virt-controller/watch/migration.go]: I refactored this a bit to hopefully make it easier to follow

File: pkg/virt-controller/watch/migration_test.go
Comment (davidvossel) [reply pkg/virt-controller/watch/migration_test.go:603]: this is a fairly common pattern we use with tables. Pass in various input and declare what output is expected. It's a way to condense more tests into a single table.

File: pkg/virt-controller/watch/migration.go
Comment (maiqueb) [reply pkg/virt-controller/watch/migration.go]: Tks. Better now.

**Review by maiqueb — APPROVED** [review]
Thanks. I think it reads better now.

**Review by kbidarkar — COMMENTED** [review]
Migration takes more than 5minutes if vm with CD-ROM.
We need to consider increasing the timeout value here, till this bug is fixed, https://bugzilla.redhat.com/show_bug.cgi?id=2014438

Both the Source and Target virt-launcher Pods would be in Running state during this 5min period.

**Comment by davidvossel** [comment]
> Migration takes more than 5minutes if vm with CD-ROM. We need to consider increasing the timeout value here, till this bug is fixed, https://bugzilla.redhat.com/show_bug.cgi?id=2014438
> 
> Both the Source and Target virt-launcher Pods would be in Running state during this long migration period of more than 5 minutes.

The timeout in this PR only get's triggered if the target pod is stuck in pending due to scheduling issues for > 5 minutes. If the target pod is running, this time out will not occur.

**Review by vladikr — COMMENTED** [review]
Looks great to me.
/approve

File: pkg/virt-controller/watch/migration.go
Comment (davidvossel) [reply pkg/virt-controller/watch/migration.go:112]: yep, we definitely could make it configurable. I'll give this some thought. I'm doing a follow up PR to this now where i have a "catch all" timeout for a target pod stuck in pending for any unknown reason, i'll need the timeout to be configurable for that simply to be able to functionally test it.

**Comment by rmohr** [comment]
> Looks great to me.
> /approve

Also for me. :+1: 

/approve

**Comment by davidvossel** [comment]
@vladikr @rmohr  take another look. I made a new change.

We now observe two timeouts
* 5m timeout will cancel any target pod stuck specifically in an "unschedulable" pending pods for > 5m
* 15m timeout will cancel any target pod stuck in pending for any reason > 15m

I created the 15m catch all to account for all the unknown reasons for why a pod will never transition successfully to a running phase, this could be issues on the node with pulling the container image, or perhaps something new we can't predict.

**Pull request merged**
